src/capture/capture.cpp

```diff
@@ -0,0 +1,268 @@
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "capture.h"
+
+#include <cerrno>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "cross.h"
+#include "fs_utils.h"
+#include "mapper.h"
+#include "render.h"
+#include "setup.h"
+#include "string_utils.h"
+#include "support.h"
+
+#include "capture_audio.h"
+#include "capture_image.h"
+#include "capture_midi.h"
+#include "capture_video.h"
+
+static std::string capturedir;
+extern const char* RunningProgram;
+uint8_t CaptureState;
+
+std::string capture_generate_filename(const char* type, const char* ext)
+{
+	if (capturedir.empty()) {
+		LOG_MSG("Please specify a capture directory");
+		return 0;
+	}
+
+	char file_start[16];
+	dir_information* dir;
+	/* Find a filename to open */
+	dir = open_directory(capturedir.c_str());
+	if (!dir) {
+		// Try creating it first
+		if (create_dir(capturedir, 0700, OK_IF_EXISTS) != 0) {
+			LOG_WARNING("Can't create dir '%s' for capturing: %s",
+			            capturedir.c_str(),
+			            safe_strerror(errno).c_str());
+			return 0;
+		}
+		dir = open_directory(capturedir.c_str());
+		if (!dir) {
+			LOG_MSG("Can't open dir %s for capturing %s",
+			        capturedir.c_str(),
+			        type);
+			return 0;
+		}
+	}
+	safe_strcpy(file_start, RunningProgram);
+	lowcase(file_start);
+	strcat(file_start, "_");
+	bool is_directory;
+	char tempname[CROSS_LEN];
+	bool testRead = read_directory_first(dir, tempname, is_directory);
+	int last      = 0;
+	for (; testRead;
+	     testRead = read_directory_next(dir, tempname, is_directory)) {
+		char* test = strstr(tempname, ext);
+		if (!test || strlen(test) != strlen(ext))
+			continue;
+		*test = 0;
+		if (strncasecmp(tempname, file_start, strlen(file_start)) != 0)
+			continue;
+		const int num = atoi(&tempname[strlen(file_start)]);
+		if (num >= last)
+			last = num + 1;
+	}
+	close_directory(dir);
+	char file_name[CROSS_LEN];
+	sprintf(file_name,
+	        "%s%c%s%03d%s",
+	        capturedir.c_str(),
+	        CROSS_FILESPLIT,
+	        file_start,
+	        last,
+	        ext);
+	return file_name;
+}
+
+FILE* CAPTURE_CreateFile(const char* type, const char* ext)
+{
+	const auto file_name = capture_generate_filename(type, ext);
+
+	FILE* handle = fopen(file_name.c_str(), "wb");
+	if (handle) {
+		LOG_MSG("Capturing %s to %s", type, file_name.c_str());
```

Comment:
Let's surround paths we print to log with quotes: `'%s'`.

Reply:
Good idea.

Reply:
Done.